source/MaterialXRuntime/Private/PrvNodeDef.cpp

```diff
@@ -12,16 +12,27 @@
 namespace MaterialX
 {
 
-PrvNodeDef::PrvNodeDef(const RtToken& name, const RtToken& nodeName) :
-    PrvValueStoringElement(RtObjType::NODEDEF, name),
+PrvNodeDef::PrvNodeDef(PrvElement* parent, const RtToken& name, const RtToken& nodeName) :
+    PrvAllocatingElement(RtObjType::NODEDEF, parent, name),
     _nodeName(nodeName),
     _numOutputs(0)
 {
 }
 
-PrvObjectHandle PrvNodeDef::createNew(const RtToken& name, const RtToken& category)
+PrvObjectHandle PrvNodeDef::createNew(PrvElement* parent, const RtToken& name, const RtToken& category)
 {
-    return std::make_shared<PrvNodeDef>(name, category);
+    if (parent && !parent->hasApi(RtApiType::STAGE))
```

Comment:
There is an edge case for this for NodeGraphs. A NodeGraph needs to have an internal node definition for it's interface, which can be dynamically updated since the graph can act as a compound where inputs and outputs are added and removed dynamically. This is an internal/private nodedef and the graph is responsible for it's life-time, keeping an explicit pointer to it. Since this is the private API I think it's OK, as we are aware and can handle the life-time explicitly in these cases. See the constructor for PrvNodeGraph, for how this is done.